sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -163,6 +164,63 @@ private[joins] object HashedRelation {
       new GeneralHashedRelation(hashTable)
     }
   }
+
+  /**
+   * Consume a `LocalNode` to create a `HashedRelation`. Because `LocalNode` is not an `Iterator`,
+   * we cannot use `apply` directly. Moreover, to avoid creating another layer of `Iterator`, we
+   * have to duplicate most codes of `apply` here.
```

Comment:
I wrote some code for this. Feel free to steal or come up with something better. (not tested!)
```
/**
 * An thin wrapper around a [[LocalNode]] that provides an iterator interface.
 */
private class LocalNodeIterator(localNode: LocalNode) extends Iterator[InternalRow] {
  private var nextRow: InternalRow = _

  override def hasNext: Boolean = {
    if (nextRow == null) {
      val res = localNode.next()
      if (res) {
        nextRow = localNode.fetch()
      }
      res
    } else {
      true
    }
  }

  override def next(): InternalRow = {
    if (hasNext) {
      val res = nextRow
      nextRow = null
      res
    } else {
      throw new NoSuchElementException
    }
  }
}
```